### DISCUSSION TITLE
add isochronous support to stm32 f3 devices

### DISCUSSION THREAD

**Discussion by cdg66** [hathach/tinyusb#836] [Q&A] [answered]
What would it need to add isochronous transfer to stm32 f3. I need it for the uac2_headset example.

**Comment by pigrew** [comment, marked as answer]
Here's a rough list of what it would take for the `stm32_fsdev` driver:

1. Modify endpoint register allocation since ISO EP require two registers.
2. Implement the double-buffering mode (since ISO EP flip flop between two buffers).
3. Likely implement DMA for mem-to-mem data copy (depending on required performance).
4. Create an API for the usb device to report buffer overrun/underruns, and CRC errors.

I think DMA would be the trickiest since it'll need to be configurable by the application (since there are only a limited number of DMA channels). I think for now all the DMA is to be blocking.

It may be also desired for the usb device driver to report CRC errors or buffer underruns to the USBD core, but I don't think there is an interface for that at the moment.

(Reviewing #593 is a good place to start, but I'm not so up to date with any of the recent isochronous work).

**Reply by cdg66** [reply]
thanks for your answer I'll use the stm32F4Discovery instead for the usb audio.  The F4 was my first choice but due to the chip shortage i had to settle for The F3 If i knew i would have use another chip. I would have love to help you to do that but im on a tight schedule. Keep writing awsome code!
